### PR TITLE
Add session conf advisor docs

### PR DESCRIPTION
**Pull request by ulysses-you**
### _Why are the changes needed?_

Add a docs for session conf advisor which includes:
1. the step of how to use
2. an example

### _How was this patch tested?_

image

### CODE REVIEW

**Comment by ulysses-you** [comment]
@yaooqinn @pan3793 @turboFei @wForget

**Review by pan3793 — APPROVED** [review]

File: docs/extensions/server/configuration.rst [hunk 1]
```diff
@@ -20,5 +20,41 @@ Inject Session Conf with Custom Config Advisor
...
-   This page is still in-progress.
+The step inject session conf with custom config advisor
+-------------------
+
+1. create a custom class which implements the `org.apache.kyuubi.plugin.SessionConfAdvisor`.
+2. adding configuration at `kyuubi-defaults.conf`:
+   ```
+   kyuubi.session.conf.advisor=${classname}
+   ```
+3. Compile and put the jar into `KYUUBI_HOME/jars`
```
Comment (pan3793) [inline docs/extensions/server/configuration.rst:31]: missing $ ?

File: docs/extensions/server/configuration.rst [hunk 2]
```diff
@@ -20,5 +20,41 @@ Inject Session Conf with Custom Config Advisor
 
 .. caution:: unstable
 
-.. warning::
-   This page is still in-progress.
+The step inject session conf with custom config advisor
```
Comment (pan3793) [inline docs/extensions/server/configuration.rst:23]: The steps of ...

File: docs/extensions/server/configuration.rst [hunk 3]
```diff
@@ -20,5 +20,41 @@ Inject Session Conf with Custom Config Advisor
...
+```
+Note that, the returned conf map will overwrite the original session conf.
+
+Example
+-------------------
+
+We have a custom class `CustomSessionConfAdvisor`:
+```java
+public class CustomSessionConfAdvisor {
+  Map<String, String> getConfOverlay(String user, Map<String, String> sessionConf) {
```
Comment (pan3793) [inline docs/extensions/server/configuration.rst:49]: `@override`

File: docs/extensions/server/configuration.rst [hunk 4]
```diff
@@ -18,7 +18,52 @@ Inject Session Conf with Custom Config Advisor
...
-.. warning::
-   This page is still in-progress.
+1. create a custom class which implements the `org.apache.kyuubi.plugin.SessionConfAdvisor`.
+2. adding configuration at `kyuubi-defaults.conf`:
+
+   .. code-block:: java
+
+      kyuubi.session.conf.advisor=${classname}
+
+3. Compile and put the jar into `$KYUUBI_HOME/jars`
```
Comment (wForget) [inline docs/extensions/server/configuration.rst:31]: Compile => compile,  keep consistent style?

**Review by wForget — APPROVED** [review]

File: docs/extensions/server/configuration.rst [hunk 5]
```diff
@@ -18,7 +18,52 @@ Inject Session Conf with Custom Config Advisor
 
 .. versionadded:: 1.5.0
 
-.. caution:: unstable
+The step of inject session conf
```
Comment (yaooqinn) [inline docs/extensions/server/configuration.rst:21]: the `steps` of `injecting` session `configs`

Comment (yaooqinn) [inline docs/extensions/server/configuration.rst:21]: Can we also elaborate a bit about what is it and what are the common use cases of it?

File: docs/extensions/server/configuration.rst [hunk 6]
```diff
@@ -18,7 +18,52 @@ Inject Session Conf with Custom Config Advisor
...
+
+   .. code-block:: java
+
+      kyuubi.session.conf.advisor=${classname}
+
+3. compile and put the jar into `$KYUUBI_HOME/jars`
+
+The `org.apache.kyuubi.plugin.SessionConfAdvisor` has a zero-arg constructor and holds one method with user and session conf and return a new conf map back.
+
+.. code-block:: java
```
Comment (yaooqinn) [inline docs/extensions/server/configuration.rst:35]: it belongs to list item 1?

File: docs/extensions/server/configuration.rst [hunk 7]
```diff
@@ -18,7 +18,52 @@ Inject Session Conf with Custom Config Advisor
...
-.. warning::
-   This page is still in-progress.
+1. create a custom class which implements the `org.apache.kyuubi.plugin.SessionConfAdvisor`.
+2. adding configuration at `kyuubi-defaults.conf`:
+
+   .. code-block:: java
+
+      kyuubi.session.conf.advisor=${classname}
+
+3. compile and put the jar into `$KYUUBI_HOME/jars`
```
Comment (yaooqinn) [inline docs/extensions/server/configuration.rst:31]: switch 2 & 3?

File: docs/extensions/server/configuration.rst [hunk 8]
```diff
@@ -18,7 +18,52 @@ Inject Session Conf with Custom Config Advisor
...
+
+.. code-block:: java
+
+   public interface SessionConfAdvisor {
+     default Map<String, String> getConfOverlay(String user, Map<String, String> sessionConf) {
+       return Collections.EMPTY_MAP;
+     }
+   }
+
+Note that, the returned conf map will overwrite the original session conf.
```
Comment (yaooqinn) [inline docs/extensions/server/configuration.rst:43]: use .. note:: ....?

File: docs/extensions/server/configuration.rst [hunk 9]
```diff
@@ -18,7 +18,52 @@ Inject Session Conf with Custom Config Advisor
...
+
+Example
+-------------------
```
Comment (yaooqinn) [inline docs/extensions/server/configuration.rst:46]:
```suggestion
-------
```

File: docs/extensions/server/configuration.rst [hunk 10]
```diff
@@ -18,7 +18,52 @@ Inject Session Conf with Custom Config Advisor
...
+     }
+   }
+
+If a user `uly` create a connection with
+
+.. code-block:: java
+
+   jdbc:hive2://localhost:10009/;hive.server2.proxy.user=uly;#spark.driver.memory=2G
+
+the final Spark application will allocate `spark.driver.memory=1G` rather 2G.
```
Comment (yaooqinn) [inline docs/extensions/server/configuration.rst:69]: `The` or `, the` final Spark application will allocate `1G` rather `than` 2G for the driver jvm.

File: docs/extensions/server/configuration.rst [hunk 11]
```diff
@@ -18,7 +18,52 @@ Inject Session Conf with Custom Config Advisor
...
+     Map<String, String> getConfOverlay(String user, Map<String, String> sessionConf) {
+       if ("uly".equals(user)) {
+         return Collections.singletonMap("spark.driver.memory", "1G");
+       } else {
+         return Collections.EMPTY_MAP;
+       }
+     }
+   }
+
+If a user `uly` create a connection with
```
Comment (yaooqinn) [inline docs/extensions/server/configuration.rst:63]: create`s`

File: docs/extensions/server/configuration.rst [hunk 12]
```diff
@@ -18,7 +18,52 @@ Inject Session Conf with Custom Config Advisor
...
+1. create a custom class which implements the `org.apache.kyuubi.plugin.SessionConfAdvisor`.
+2. adding configuration at `kyuubi-defaults.conf`:
+
+   .. code-block:: java
+
+      kyuubi.session.conf.advisor=${classname}
+
+3. compile and put the jar into `$KYUUBI_HOME/jars`
+
+The `org.apache.kyuubi.plugin.SessionConfAdvisor` has a zero-arg constructor and holds one method with user and session conf and return a new conf map back.
```
Comment (yaooqinn) [inline docs/extensions/server/configuration.rst:33]: `returns`

Comment (yaooqinn) [reply docs/extensions/server/configuration.rst:33]: `and` holds -> `,` holds

Comment (yaooqinn) [reply docs/extensions/server/configuration.rst:33]: returns has the back meaning inside the word, we can remove the last redundant `back`.

**Comment by yaooqinn** [comment]
`` means italic in .rst, if you want to highlight like .md, then ```` is what you need

**Comment by ulysses-you** [comment]
>  `` means italic in .rst, if you want to highlight like .md, then ```` is what you need

yes, that is what I need

**Comment by turboFei** [comment]
Flaky test:
```
KyuubiOperationKubernetesClusterClusterModeSuite:
- Spark Cluster Mode On Kubernetes Kyuubi KubernetesApplicationOperation Suite *** FAILED ***
  failKillResponse._1 was true (SparkOnKubernetesTestsSuite.scala:213)
```

File: docs/extensions/server/configuration.rst [hunk 13]
```diff
@@ -18,7 +18,56 @@ Inject Session Conf with Custom Config Advisor
 
 .. versionadded:: 1.5.0
 
-.. caution:: unstable
+Session Conf Advisor
+--------------------
 
-.. warning::
-   This page is still in-progress.
+Kyuubi supports inject session configs with custom config advisor. It is usually used to append or overwrite session configs dynamically, so administrators of Kyuubi can have an ability to control the user specified configs.
```
Comment (turboFei) [inline docs/extensions/server/configuration.rst:24]: nit: custom -> customized

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
https://github.com/apache/incubator-kyuubi/issues/3100 please link it in the Umbrella.

**Comment by ulysses-you** [comment]
thank you for review, merging to master

**Pull request closed without merging**